tools/publish-crates.sh

```diff
@@ -4,41 +4,78 @@ set -euo pipefail
 
 DRY_RUN=0
+ALLOW_DIRTY=0
+NEW_CRATE_OWNERS=("tyler@clockworklabs.io" "zeke@clockworklabs.io")
 
-if [ "$#" != "0" ]; then
-    if [ "$1" != "--dry-run" ]; then
-        echo "$1 is not a valid flag";
-        exit 1;
-    else
-        DRY_RUN=1
-    fi
-fi
+# Parse arguments
+while [[ $# -gt 0 ]]; do
+    case "$1" in
+        --dry-run)
+            DRY_RUN=1
+            ;;
+        --allow-dirty)
+            ALLOW_DIRTY=1
+            ;;
+        *)
+            echo "Invalid argument: $1"
+            exit 1
+            ;;
+    esac
+    shift
+done
 
-if [ $DRY_RUN != 1 ] ; then
-	echo "You are about to publish to crates.io (dry run is false.)"
-	echo "We are also going to do a test install after publishing. This will remove any version of spacetimedb-cli you may have installed."
-	echo
-	echo "Press [Enter] to continue."
-	read -r
+if [ $DRY_RUN -ne 1 ]; then
+    echo "You are about to publish to crates.io (dry run is false.)"
+    echo "We are also going to do a test install after publishing. This will remove any version of spacetimedb-cli you may have installed."
+    echo
+    echo "Press [Enter] to continue."
+    read -r
 fi
 
 BASEDIR=$(pwd)
-declare -a CRATES=("metrics" "primitives" "sql-parser" "bindings-macro" "bindings-sys" "data-structures" "sats" "lib" "schema" "bindings" "table" "vm" "client-api-messages" "paths" "commitlog" "durability" "fs-utils" "snapshot" "expr" "core" "client-api" "standalone" "cli" "sdk")
+declare -a CRATES=("metrics" "primitives" "sql-parser" "bindings-macro" "bindings-sys" "data-structures" "sats" "lib" "schema" "bindings" "table" "vm" "client-api-messages" "paths" "commitlog" "durability" "fs-utils" "snapshot" "expr" "execution" "physical-plan" "query" "core" "client-api" "standalone" "cli" "sdk")
 
-for crate in "${CRATES[@]}" ; do
-	if [ ! -d "${BASEDIR}/crates/${crate}" ] ; then
-		echo "This crate does not exist: ${crate}"
-		exit 1
-	fi
+for crate in "${CRATES[@]}"; do
+    if [ ! -d "${BASEDIR}/crates/${crate}" ]; then
+        echo "This crate does not exist: ${crate}"
+        exit 1
+    fi
 done
 
-for crate in "${CRATES[@]}" ; do
-	cd "${BASEDIR}/crates/${crate}"
-	if [ $DRY_RUN == 1 ] ; then
-		cargo publish --dry-run
-	else
-		cargo publish
-	fi
+for crate in "${CRATES[@]}"; do
+    cd "${BASEDIR}/crates/${crate}"
+
+    PUBLISH_CMD="cargo publish"
+    [[ $DRY_RUN -eq 1 ]] && PUBLISH_CMD+=" --dry-run"
+    [[ $ALLOW_DIRTY -eq 1 ]] && PUBLISH_CMD+=" --allow-dirty"
+
+    # Check if crate exists on crates.io
+    if ! cargo search "$crate" --limit 1 | grep -q "^$crate ="; then
+        IS_NEW_CRATE=1
+        echo "INFO: Detected $crate as a new crate on crates.io!"
+    else
+        IS_NEW_CRATE=0
+    fi
+
+    echo "Publishing crate: $crate with command: $PUBLISH_CMD"
+    if ! OUTPUT=$($PUBLISH_CMD 2>&1); then
+        if echo "$OUTPUT" | grep -q "crate version .* is already uploaded"; then
+            echo "WARNING: Crate $crate version is already published. Skipping..."
```

Comment:
as I've mentioned before, I think we need to make this configurable with a command-line option. This could be a genuine failure if e.g. we failed to bump package versions entirely.

Can we add e.g. `--skip-already-published`?

Reply:
(I can make this change if you're busy, it's easy now that you've refactored our argument parsing)